### PR TITLE
add lookahead optimiser

### PR DESCRIPTION
**Pull request by chengchingwen**
This add the implementation of Lookahead optimiser from [this paper](https://arxiv.org/abs/1907.08610) and based on the [pytorch impl](https://github.com/lonePatient/lookahead_pytorch/blob/master/optimizer.py)

### CODE REVIEW

File: src/optimise/Optimise.jl [hunk 1]
```diff
@@ -2,10 +2,12 @@ module Optimise
 
 export train!,
 	SGD, Descent, ADAM, Momentum, Nesterov, RMSProp,
-	ADAGrad, AdaMax, ADADelta, AMSGrad, NADAM, ADAMW,RADAM, 
-	InvDecay, ExpDecay, WeightDecay, stop, Optimiser
+	ADAGrad, AdaMax, ADADelta, AMSGrad, NADAM, ADAMW,
+  RADAM, Lookahead, InvDecay, ExpDecay, WeightDecay,
```
Comment (bhvieira) [inline src/optimise/Optimise.jl]: The indentation here (and the next line) moved. Just letting you know.

Comment (chengchingwen) [reply src/optimise/Optimise.jl]: There miss a space between `ADAMW` and `RADAM` but with the space that line looks too long, so I moved it to the next line.

Comment (bhvieira) [reply src/optimise/Optimise.jl]: Sure, but the indentation is different now. You could make every line use a double-space as well, or move back to tab everywhere, I'm not sure Flux follow any single style.

Comment (chengchingwen) [reply src/optimise/Optimise.jl]: what do you mean? all lines are two spaces indent.

Comment (bhvieira) [reply src/optimise/Optimise.jl]: Weird, my browser renders it as if the line starting with RADAM was a double-space while the other one above was a tab.

Comment (chengchingwen) [reply src/optimise/Optimise.jl]: Ah, I see. I set all tab as double-space width.

File: src/optimise/lookahead.jl [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+"""
+mutable struct Lookahead{O, MH <: Function}
+  opt::O
+  momentum_handler::MH
+  alpha::Float64
+  k::Int
+  state::IdDict
+end
+
+const MomentumOptim = Union{Momentum, RMSProp, Nesterov, ADAM, RADAM, AdaMax, ADAGrad, ADADelta, AMSGrad, NADAM}
```
Comment (dhairyagandhi96) [inline src/optimise/lookahead.jl]: Perhaps an `AbstractOptimiser` type would help with some of the repitition.

Comment (chengchingwen) [reply src/optimise/lookahead.jl]: But some optimiser don't have momentum term. Maybe we also need a `AbstractMomentumOptimiser`?

Comment (dhairyagandhi96) [reply src/optimise/lookahead.jl]: Thinking if we could use traits here?

Comment (chengchingwen) [reply src/optimise/lookahead.jl]: what's that?

Comment (chengchingwen) [reply src/optimise/lookahead.jl]:
```julia
abstract type OptimiserStyle end
struct Momentumlike <: OptimiserStyle end
struct NonMomentumlike <: OptimiserStyle end

has_momentum(o::Type{Descent}) = NonMomentumlike()
has_momentum(o::Type{Momentum}) = Momentumlike()
has_momentum(o::Type{Adam}) = Momentumlike()
```
like this?

Comment (dhairyagandhi96) [reply src/optimise/lookahead.jl]: Something like that, yeah.

Comment (dhairyagandhi96) [reply src/optimise/lookahead.jl]: Also, ref https://github.com/oxinabox/oxinabox.github.io/blob/master/notebook_posts/Dispatch%2C%20Traits%20and%20Metaprogramming%20Over%20Reflection.ipynb

**Comment by chengchingwen** [comment]
@dhairyagandhi96  I define a `AbstractOptimiser`and the `GradientStyle`traits`.

**Comment by CarloLucibello** [comment]
I would step back to the original problem and just add the Lookahead optimizer

**Comment by chengchingwen** [comment]
@CarloLucibello  Sure! should I revert the commit to 510cfaa or 14956c2 ?

**Comment by CarloLucibello** [comment]
yes please. If we want to introduce a type hierarchy for the optimizers let's first discuss it in a separate issue/PR

**Comment by chengchingwen** [comment]
@CarloLucibello Reverted.

**Comment by CarloLucibello** [comment]
In the paper, I read 
"""
We have the choice of maintaining, interpolating, or resetting the internal state (e.g. momentum) of
the inner optimizer. We evaluate this tradeoff on the CIFAR dataset (where every choice improves
convergence) in Appendix D.1 and maintain internal state for the other experiments.
....
D.1 Inner Optimizer State
Throughout our paper, we maintain the state of our inner optimizer for simplicity. For SGD with
heavy-ball momentum, this corresponds to preserving the momentum. Here, we present a sensitivity
study by comparing the convergence of Lookahead when maintaining the momentum, interpolating
the momentum, and resetting the momentum. All three improve convergence versus SGD.
"""
So let's keep this simple and maintain the state.

This also needs a rebase

**Comment by CarloLucibello** [comment]
> So let's keep this simple and just maintain the state.

I'm saying this because the specific needs of this PR, we probably need a state setter/getter API for the optimizers, so we should think of it more carefully in a separate issue/PR, then we can revisit LookAhead when we have that in place.
So I think it's better to just merge the minimal LookAhead implementation here, with no state reset

**Comment by chengchingwen** [comment]
@DhairyaLGandhi rebased and updated

**Comment by CarloLucibello** [comment]
sorry we droppebd the ball here. Closing since optimisers are now in Optimisers.jl

**Pull request closed without merging**